scripts/nlp_language_modeling/convert_nemo_gpt_to_mcore.py

```diff
@@ -0,0 +1,228 @@
+    if skip_if_output_exists and os.path.exists(output_ckpt_file):
+        logging.info(f"Output file already exists ({output_ckpt_file}), skipping conversion...")
+        return
+    dummy_trainer = Trainer(devices=1, accelerator='cpu')
+
+    nemo_model = MegatronGPTModel.restore_from(input_ckpt_file, trainer=dummy_trainer)
+    nemo_tokenizer_model = nemo_model.cfg.tokenizer.model
+    nemo_state_dict = nemo_model.state_dict()
+    mcore_state_dict = OrderedDict()
+    for mcore_param, nemo_param in build_key_mapping(nemo_model.cfg).items():
+        mcore_state_dict[mcore_param] = nemo_state_dict[nemo_param]
+
+    mcore_model = get_mcore_model_from_nemo_ckpt(input_ckpt_file)
+    mcore_model = load_model(mcore_model, mcore_state_dict)
+
+    if nemo_model.cfg.tokenizer.model is not None:
+        logging.info("registering artifact: tokenizer.model = " + nemo_tokenizer_model)
+        mcore_model.register_artifact("tokenizer.model", nemo_tokenizer_model)
+
+    mcore_model.save_to(output_ckpt_file)
+    logging.info(f"Done. Model saved to {output_ckpt_file}")
+
+
+def run_sanity_checks(nemo_ckpt_file, mcore_ckpt_file):
+    cfg = OmegaConf.load(
+        os.path.join(
+            os.path.dirname(__file__),
+            '../../examples/nlp/language_modeling/tuning/conf/megatron_gpt_peft_tuning_config.yaml',
+        )
+    )
+
+    cfg.trainer.precision = 'bf16'  # change me
+    dtype = torch.bfloat16
+    trainer = MegatronTrainerBuilder(cfg).create_trainer()
+    nemo_model = MegatronGPTModel.restore_from(nemo_ckpt_file, trainer=trainer).eval().to(dtype)
+    mcore_model = MegatronGPTModel.restore_from(mcore_ckpt_file, trainer=trainer).eval().to(dtype)
+
+    logging.debug("*** Mcore model restored config")
+    logging.debug(OmegaConf.to_yaml(mcore_model.cfg))
+
+    nemo_summary = nemo_model.summarize()
+    mcore_summary = mcore_model.summarize()
+
+    logging.info("Sanity checks:")
+
+    # check num weights match
+    assert nemo_summary.total_parameters == mcore_summary.total_parameters, "❌ total parameters do not match"
+    assert nemo_summary.model_size == mcore_summary.model_size, "❌ model sizes do not match"
+    logging.info("✅ Number of weights match")
+
+    # check weights match
+    mcore_state_dict = mcore_model.state_dict()
+    nemo_state_dict = nemo_model.state_dict()
+    for mcore_param, nemo_param in build_key_mapping(nemo_model.cfg, use_O2_prefix=False).items():
+        try:
+            assert torch.allclose(
+                mcore_state_dict[mcore_param], nemo_state_dict[nemo_param]
+            ), f"❌ parameter {mcore_param} does not match"
+        except KeyError:
+            buffers = [k for k, v in mcore_model.named_buffers()]
+            assert (
+                mcore_param in buffers or mcore_param.replace('model.', 'model.module.', 1) in buffers
+            ), f"❌ parameter {mcore_param} is not found in the state dict or named_buffers()"
+    logging.info("✅ Weights match")
```

Comment:
what happens in the case where there is a new item in mcore's state dict? since we only verify items that are in `build_key_mapping` we'll skip any extra items in either of mcore/nemo state dict?
during sanity check can we somehow check that we have looked into all params and are not missing anything in we have not missed any param that exists in the state_dict but is not in `build_key_mapping` ?

Reply:
Sounds good, I'll add another test to make sure there's no extraneous params